text/0225-configuration.md

```diff
@@ -0,0 +1,329 @@
+* Broadly supported format. Ideally, the information encoded in the file can be decoded using native tools for all OpenTelemetry implementation languages. However, it must be possible for languages that do not natively support an encoding format to write their own parsers.
+* The configuration format must support structured data. At the minimum arrays and associative arrays.
+* The format must support at least null, boolean, string, double precision floating point (IEEE 754-1985), and signed 64 bit integer value types.
+* Custom span processors, exporters, samplers, or other user defined extension components can be configured using this format.
+* Configure SDK, but also configure instrumentation.
+* Must offer stability guarantees while supporting evolution.
+* The structure of the configuration can be validated via a schema.
+* Support environment variable substitution to give users the option to avoid storing secrets in these files.
+
+## Internal details
+
+The schema for OpenTelemetry configuration is to be published in a repository to allow language implementations to leverage that definition to automatically generate code and/or validate end-user configuration. This will ensure that all implementations provide a consistent experience for any version of the schema they support. An example of such a proposed schema is available [here](https://github.com/MrAlias/otel-schema/tree/main/json_schema/schema).
+
+The working group proposes the use of [JSON Schema](https://json-schema.org/) as the language to define the schema. It provides:
+
+* support for client-side validation
+* code generation
+* broad support across languages
+
+In order to provide a minimal API surface area, implementations *MUST* support the following methods.
+
+### Configure(config)
+
+An API called `Configure` receives a configuration object. This method then applies the configuration object's details to the SDK. This method specifically applies the configuration object to allow for multiple configuration format providers to be supported in the future. This OTEP describes two such providers in a file and data structure formats below, but remote file formats *MAY* be implemented in the future.
+
+### ParseAndValidateConfigurationFromFile(filepath, format) -> config
+
+An API called `ParseAndValidateConfigurationFromFile` receives a string parameter indicating the file path containing the configuration to be parsed. An optional format parameter may be provided to indicate the format that this configuration uses. YAML *SHOULD* be supported, but JSON, which is a strict subset of YAML, *MAY* be supported if a language ecosystem has good reason. For example a language that has built-in support for JSON running in an environment that is sensitive to additional dependencies *MAY* only be able to support JSON configuration.
```

Comment:
I've spent quite a bit of time thinking about this paragraph. I'm not sure if it needs to be changed or not, but I'll write my thoughts here and we'll see what the authors think.

"Sensitive to additional dependencies" seems to refer to JS or even more specifically web JS. I'm not sure file-based configuration as it is specified here (loaded from a file path) makes sense in the web context. It might make sense to distinguish between implementations which ship their own dependencies (most compiled languages) and environments which don't (most dynamic languages). Taking NodeJS as an example, most users are not overly sensitive to dependencies, in fact many are unfortunately desensitized in this area, but they introduce complications that many people don't think about. As JS maintainer I want to limit or eliminate dependencies whenever possible for more reasons than just deployment size.

JS has built-in support for JSON, Ruby has standard library support for both YAML (as of version 2) and JSON (as of version 3), and Python has standard library support for JSON. I think it makes sense to write this in such a way that it encourages whatever format can be used without any external dependencies, preferring YAML if the state of support is equal for both.

Maybe something along these lines?

```suggestion
An API called `ParseAndValidateConfigurationFromFile` receives a string parameter indicating the file path containing the configuration to be parsed. An optional format parameter may be provided to indicate the format that this configuration uses. At least one of JSON or YAML MUST be supported. If either format can be supported without additional dependencies, that format SHOULD be preferred. If neither or both formats are supported, YAML should be the preferred choice.
```

Reply:
The question of the format stems from poor composition. I see several layers here:
  * some sort of a factory mechanism that takes a "configuration" and produces SDK components
  * a "configuration" data model. Agnostic to how it is instantiated.
  * a configuration parser, for a certain format / representation.

```
cfg = ConfigYAMLParser.parse(File.load(filename))
factory = SDKFactory.fromCfg(cfg)
tracerProvider = factory.tracerProvider()
```
This is visually more complicated, but cleanly decomposed by responsibilities. I wouldn't prefer `ParseAndValidateConfigurationFromFile` over above.

Reply:
> I think it makes sense to write this in such a way that it encourages whatever format can be used without any external dependencies, preferring YAML if the state of support is equal for both.

The downside of not having a stronger preference for one format is that it forces users to have to be aware of the format preference of their language. If all otel languages support YAML, users can more easily use standard config across all languages in use by their organization.

Reply:
Is there any feature of YAML that we're specifically using aside from a preference for not using curly braces? If there is a desire to use a single configuration across all services, JSON is actually valid YAML so I would argue that would be a better choice. A valid JSON file can be used with OTel services which support either format. Introducing spec-required external dependencies is something I think we should consider very carefully before moving forward.

Reply:
> The question of the format stems from poor composition. I see several layers here:
> 
> * some sort of a factory mechanism that takes a "configuration" and produces SDK components
> * a "configuration" data model. Agnostic to how it is instantiated.
> * a configuration parser, for a certain format / representation.
> 
> ```
> cfg = ConfigYAMLParser.parse(File.load(filename))
> factory = SDKFactory.fromCfg(cfg)
> tracerProvider = factory.tracerProvider()
> ```
> 
> This is visually more complicated, but cleanly decomposed by responsibilities. I wouldn't prefer `ParseAndValidateConfigurationFromFile` over above.

I think the motivation for `ParseAndValidateConfigurationFromFile` which mixes many concerns is to support the `OTEL_CONFIG_FILE` variable. Not arguing for that function, just seems like that was the motivation.

Reply:
> Is there any feature of YAML that we're specifically using aside from a preference for not using curly braces? If there is a desire to use a single configuration across all services, JSON is actually valid YAML so I would argue that would be a better choice.

In my opinion, the reasons for YAML over JSON are:

* YAML is already used for configuring the collector
* support for comments
* more human readable than JSON

Reply:
I wasn't questioning the choice to prefer YAML. Just pointing out that if a user has multiple services and one of them only supports JSON, they can still use a single config file for both because all valid JSON is valid YAML.

Reply:
Also, YAML supports anchors, which allow reusing bits of config.

In addition to the lack of curly braces, YAML also removes the need for double quotes. Generally, I think YAML is easier for humans to write than JSON.

Reply:
> In my opinion, the reasons for YAML over JSON are:
> 
> * YAML is already used for configuring the collector

Since the config between collector and SDKs is unlikely to be shared, I think this advantage is quite minor

> * support for comments

Comments are nice I agree.

> * more human readable than JSON

IMO this really only applies to badly-formatted JSON

Reply:
> Generally, I think YAML is easier for humans to write than JSON.

Maybe its my JS background but I strongly disagree. Users not already familiar with YAML have deal with data type ambiguity, significant whitespace, multiple list syntaxes, weird syntax with nested list/objects, etc.

In any case, I'm not arguing for JSON over YAML. I'm arguing for reduction of external dependencies and conformance with the existing ecosystems, which may entail choosing JSON in some languages.

Reply:
One possible solution is to allow users to optionally install pluggable parsers. For example in JS we could use JSON by default, but parse additional filetypes if their parser exists:

```sh
$ OTEL_CONFIG_FILE=./otel.yaml npm start
Error: no parser found for configuration file with extension 'yaml'
       Use a supported file type or install a parser for this file type

$ npm install @opentelemetry/config-parser-yaml
success

$ OTEL_CONFIG_FILE=./otel.yaml npm start
process starting
```

Reply:
This also opens the door to additional file types in the future (another very common idiom in JS is to actually use `.js` files instead of `.json` files for configuration.

Reply:
@jack-berg I saw you thumbed up my message but i'm not sure if you're saying that would be allowable by the current wording, or if you think it would be a desirable outcome in general. I'm hoping for the second one. Maybe you'd be happier with something like this:

```suggestion
An API called `ParseAndValidateConfigurationFromFile` receives a string parameter indicating the file path containing the configuration to be parsed. An optional format parameter may be provided to indicate the format that this configuration uses. At least one of JSON or YAML MUST be supported. If either format can be supported without additional dependencies, that format SHOULD be preferred. If neither or both formats are supported, YAML should be the preferred choice. If YAML is not supported due to dependency concerns, there SHOULD be a way for a user to explicitly enable it by installing their own dependency.
```

Reply:
> or if you think it would be a desirable outcome in general. 

This. I like your wording, but would weaken "there SHOULD be a way for a user to explicitly enable it by installing their own dependency" to a "MAY". Hoping to minimize burden on maintainers and promote standardization where possible. If a language community gets enough feedback that the defaults are insufficient, let them opt into adding this extension capability.

Reply:
Adding suggestion with MAY wording for quick application if @codeboten agrees

```suggestion
An API called `ParseAndValidateConfigurationFromFile` receives a string parameter indicating the file path containing the configuration to be parsed. An optional format parameter may be provided to indicate the format that this configuration uses. At least one of JSON or YAML MUST be supported. If either format can be supported without additional dependencies, that format SHOULD be preferred. If neither or both formats are supported, YAML should be the preferred choice. If YAML is not supported due to dependency concerns, there MAY be a way for a user to explicitly enable it by installing their own dependency.
```

Reply:
Thanks @dyladan, this works for me